tests/dsp_testing/filters.ipynb

```diff
@@ -0,0 +1,947 @@
+    "            title = f'Filtered Rawrf Samples - Stage {i}'\n",
+    "            plot_filter_stage(input_samples, antenna_output, filter[0, :],\n",
+    "                              scheme.stages[i].input_rate, scheme.stages[i].output_rate,\n",
+    "                              title, center_freq=mixing_freqs[0], freq_lim=freq_lim)\n",
+    "        else:\n",
+    "            antenna_input = output_samples[i-1][0, 0, :]\n",
+    "            title = f'Filtered Rawrf Samples - Stage {i}'\n",
+    "            plot_filter_stage(antenna_input, antenna_output, filter[0, :],\n",
+    "                              scheme.stages[i].input_rate, scheme.stages[i].output_rate,\n",
+    "                              title, freq_lim=freq_lim)\n",
+    "\n",
+    "mixing_freq = [-1.5e6]\n",
+    "sequence_samples = sequence_samples_cpu[np.newaxis, :]\n",
+    "default_outputs, mixed_filters = mock_dsp(sequence_samples, mixing_freq, default_scheme)\n",
+    "\n",
+    "# plot the results\n",
+    "plot_dsp_process(default_scheme, sequence_samples[0], default_outputs, mixed_filters, mixing_freq)\n",
+    "# plot_dsp_process(default_scheme, sequence_samples[0], default_outputs, mixed_filters, mixing_freq, zoom_in=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "878c6ad3-88a4-4d63-bf9d-51572d034aa6",
+   "metadata": {},
+   "source": [
+    "## New filter usage on real data\n",
+    "Now, we can compare with the performance of the new filter on the real data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "75f31222-6e39-4f47-b6e4-712800d84be5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "new_outputs, new_mixed_filters = mock_dsp(sequence_samples, mixing_freq, new_filter)\n",
+    "\n",
+    "# plot the results\n",
+    "plot_dsp_process(new_filter, sequence_samples[0], new_outputs, new_mixed_filters, mixing_freq)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5179b5c6-a8b2-448f-af48-5b5038b11170",
+   "metadata": {},
+   "source": [
+    "## Direct comparison of filter outputs\n",
+    "Now, we plot the output of both the default filter scheme and the new filter scheme, to compare how the resulting antennas_iq data looks."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "60bd94b2-73ca-4bd2-824e-a82584df3fb5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "final_data_1 = default_outputs[-1]\n",
+    "final_data_2 = new_outputs[-1]\n",
+    "\n",
+    "antenna_data_1 = final_data_1[0, 0, :]\n",
+    "antenna_data_2 = final_data_2[0, 0, :]\n",
+    "legend = ['Default Scheme', 'New Scheme']\n",
+    "title = f'Filter Scheme Comparison'\n",
+    "_ = plot_freq_time_comparison(antenna_data_1, antenna_data_2, default_scheme.output_rates[-1], title, legend)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "32b882fe-fcf7-4f1b-a526-1ae7a9c3dd7b",
+   "metadata": {},
+   "source": [
+    "## Benchmark filter performance\n",
```

Comment:
I cannot test this as my computer doesn't have a GPU and the process hangs forever when I run it. I'll assume it is good.